### PR TITLE
Storing pre-prepares in KV and ledger replay adjustments

### LINKED ISSUE #539: PBFT to store meta data in the KV that will be in turn encrypted, serialized and persisted to the ledger
**Overview**

Right now PBFT uses a `LedgerEnclave` object to write pre-prepares and requests to the ledger. These are not encrypted. We want PBFT to store this "meta data" in the KV and from there eventually persist to the ledger.

We want two types of KV maps. `derived` maps and `replicated` maps. We want to encrypt and serialize both, but we don't want to replicate the derived maps or persist them to the ledger.

In Raft all maps (at least for now) are considered replicated ones.

In the PBFT scenario where we have derived maps and replicated maps, we want to implement two Merkle trees. One over the replicated maps only and one over the entire state of the KV (replicated and derived maps). Both types of maps need to therefore be serialized and encrypted.

**Action points**

From the CCF side the steps that need to be taken are:

- [x] 1. Add a property (Boolean) to the KV `Map` that indicates if we want that map to replicated or not, defined on map creation based on the consensus protocol being built
- [x] 2. `Tx::serialise()` needs to return two vectors, one containing the derived maps write set encrypted and serialized and one containing the replicated maps write set encrypted and serialized
- [x] 3. Create the map where the pre-prepare and its request(s) will be stored into
- [x] 4. Create two merkle trees, one being updated due to the replicated maps and one being updated due to all of the maps

From the PBFT side of things the steps that need to be taken are:

- [x] 5. The flow of a request batch will need to change. 
   - a. A pre-prepare gatheres enough requests and is ready to execute them (call to `execute_tentative`
   - b. Every call to `process_pbft` will create a new transaction, write the request to the replicated request table, populate the derived tables by executing the request and then commit (and finally will trigger replicate)
  -  [x] c. This will update the two merkle trees and return to `Replica` the merkle tree roots (as is done now but for one merkle tree)
  - [x] d. When the requests are done, pre-prepare can be completed by having the two merkle roots written to it. 
  - e. Replica then calls `process_pbft` again with a "request" to store that pre-prepare in the replicated request table, which will in turn be written to the ledger
  - f. The pre-prepare can be broadcasted to the backups as usual

**Points of discussion**

Regarding what happens when `Consensus::replicate()` is called in the PBFT scenario.

Right now, PBFT uses its own channels to distribute requests and pre-prepares. Requests get sent via the client-proxy and pre-prepares (along with any other pbft-related message) via the Node's send() method.

We can either implement `Consensus::replicate()` to only persist the entries to the ledger and keep the rest of the PBFT communication as it is now.

OR

We can implement `Consensus::replicate()` to actually replicate the tables that hold the requests and the pre-prepares. This has the nice effect that the two consensuses behave in a similar way. But then we need to have client-proxy not send the requests (which is an easy thing to do) but we also need to tell PBFT to not send out the pre-prepares as they will be communicated via the table replication. This has the down side of creating two ways for PBFT replicas to communicate, one via the protocol (which they still need to do for the prepare, commit, response messages) and one via `replicate` for the requests and the pre-prepares. We could have all of the messages written to the KV and communicated via `replicate`, but then we will need to write everything to the KV tables and persist them to the ledger which I don't think is something that we want to do.

### PR DESCRIPTION
**Pull request by olgavrou**
Resolves #539

Storing pre-prepares in kv

- `pbft_info` table to store all pbft-related data (requests and pre-prepares for now)
- `LedgerWriter` takes `Store` and `pbft_info` so that it can store pre-prepares into the kv (it will later be able to store view changes/prepare proof)
- ledger replay needed adjustment as we now don't manually write requests and pre-prepares to the ledger, but let the data be written to the ledger after `replicate()`
- flatbuffers deserialiser to ignore derived tables when running with pbft
- there is a playback hook on the `pbft_info` table. When data will be read from the ledger, it needs to be deserialised via `Store::deserialise`. This will trigger the local commit hook and call `apply_ledger_data` in `Replica.cpp` that can then execute pre-prepares when appropriate. This hook will need to be turned on when we get replay messages and turned off again for normal execution as we do not want the commit hook to be triggered when we add pre-prepares and requests to the kv in normal execution

Feedback much wanted :)

### CODE REVIEW

**Comment by cimetrics** [comment]
store_pre_prepares@[3554 aka 20191218.20](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3554&view=results) vs master ewma over 20 builds from [3327](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3327&view=results) to [3546](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3546&view=results)
images

File: src/consensus/pbft/libbyz/LedgerWriter.cpp [hunk 1]
```diff
@@ -42,73 +43,29 @@ void LedgerWriter::write_prepare(
...
-      else
-      {
-        Pre_prepare_ledger_large_message_header header(0);
-        serialized::write(data, entry_size, (uint8_t*)&header, sizeof(header));
-      }
-    }
-  }
-  ledger->put_entry(entry);
+  auto version = store->next_version();
+  LOG_TRACE << "Storing pre prepare at seqno " << pp->seqno() << std::endl;
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/LedgerWriter.cpp]: LOG_TRACE_FMT("Storing pre-prepare at seqno {}", pp->seqno())

File: src/consensus/pbft/libbyz/LedgerWriter.cpp [hunk 2]
```diff
@@ -42,73 +43,29 @@ void LedgerWriter::write_prepare(
...
+  store->commit(
+    version,
+    [version, pp, this]() {
+      ccf::Store::Tx tx(version);
+      auto info_view = tx.get_view(*pbft_info);
+      info_view->put(
+        0,
+        {pbft::InfoType::PRE_PREPARE,
+         {pp->seqno(),
+          pp->size(),
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/LedgerWriter.cpp]: Why do we store the size?

File: src/consensus/pbft/libbyz/LedgerWriter.cpp [hunk 3]
```diff
@@ -42,73 +43,29 @@ void LedgerWriter::write_prepare(
...
+          pp->num_big_reqs(),
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/LedgerWriter.cpp]: Need to add the digest signature here too.

File: src/consensus/pbft/libbyz/LedgerWriter.cpp [hunk 4]
```diff
@@ -4,8 +4,9 @@
 
 #include "Request.h"
 
-LedgerWriter::LedgerWriter(std::unique_ptr<consensus::LedgerEnclave> ledger_) :
-  ledger(std::move(ledger_))
+LedgerWriter::LedgerWriter(pbft::Store* store_, pbft::PbftInfo* pbft_info_) :
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/LedgerWriter.cpp]: Refs of shared_ptr

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp [hunk 5]
```diff
@@ -77,8 +75,7 @@ NodeInfo get_node_info()
 }
 
 void create_replica(
-  std::vector<char>& service_mem,
-  std::unique_ptr<consensus::LedgerEnclave> ledger)
+  std::vector<char>& service_mem, pbft::Store* store, pbft::PbftInfo* pbft_info)
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: No raw pointers please.

File: src/consensus/pbft/pbftinfo.h [hunk 6]
```diff
@@ -63,11 +80,37 @@ namespace pbft
...
-  // them into the ledger
-  using PbftRequests = ccf::Store::Map<size_t, Request>;
-
   DECLARE_JSON_TYPE(Request);
   DECLARE_JSON_REQUIRED_FIELDS(Request, actor, caller_id, caller_cert, raw);
+
+  struct PrePrepare
+  {
+    int64_t seqno;
+    int message_size;
```
Comment (achamayou) [inline src/consensus/pbft/pbftinfo.h]: size_t

File: src/consensus/pbft/pbftinfo.h [hunk 7]
```diff
@@ -63,11 +80,37 @@ namespace pbft
...
+    int16_t num_big_requests;
```
Comment (achamayou) [inline src/consensus/pbft/pbftinfo.h]: uint16_t

File: src/consensus/pbft/pbfttables.h [hunk 8]
```diff
@@ -8,10 +8,10 @@ namespace pbft
 {
   struct Tables
   {
-    static constexpr auto PBFT_REQUESTS = "ccf.pbft.requests";
+    static constexpr auto PBFT_INFO = "ccf.pbft.info";
```
Comment (achamayou) [inline src/consensus/pbft/pbfttables.h]: Let's have prepares and requests in separate tables with separate hooks please.

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 9]
```diff
@@ -1045,6 +1064,17 @@ void Replica::register_global_commit(global_commit_handler_cb cb, void* ctx)
   global_commit_ctx = ctx;
 }
 
+template <class T>
+std::unique_ptr<T> Replica::create_message(
+  const uint8_t* message_data, size_t data_size)
+{
+  Message* m = new Message(Max_message_size);
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: you do not to alloc max_message_size as you know the type you are converting to.

File: src/consensus/pbft/libbyz/parameters.h [hunk 10]
```diff
@@ -7,6 +7,13 @@
 
 #include <cstddef>
 
+// keypair parameters
```
Comment (olgavrou) [inline src/consensus/pbft/libbyz/parameters.h]: moved here so we can use `signature_size` in `pbftpreprepares.h`

File: src/ds/json_schema.h [hunk 11]
```diff
@@ -51,6 +51,9 @@ namespace ds
     template <typename T>
     void fill_schema(nlohmann::json& schema);
 
+    template <typename T>
```
Comment (olgavrou) [inline src/ds/json_schema.h]: needed here, same as `fill_schema`, otherwise we will get `undeclared identifier` compiler error since using `std::array` for storing the `digest_sig` in `pbftpreprepares.h` will call `fill_json_schema`

**Comment by cimetrics** [comment]
store_pre_prepares@[3597 aka 20191220.16](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3597&view=results) vs master ewma over 20 builds from [3376](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3376&view=results) to [3589](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3589&view=results)
images

File: src/consensus/pbft/libbyz/parameters.h (hunk 10 above)
Comment (achamayou) [reply src/consensus/pbft/libbyz/parameters.h]: What do we use SHA512 for?

**Comment by cimetrics** [comment]
store_pre_prepares@[3601 aka 20191220.18](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3601&view=results) vs master ewma over 20 builds from [3392](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3392&view=results) to [3599](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3599&view=results)
images

File: src/consensus/pbft/libbyz/test/replica_test.cpp [hunk 12]
```diff
@@ -376,6 +377,14 @@ int main(int argc, char** argv)
...
   }
 
+  auto store = std::make_shared<ccf::Store>(
+    pbft::replicate_type_pbft, pbft::replicated_tables_pbft);
+  auto& pbft_requests = store->create<pbft::Requests>(
+    pbft::Tables::PBFT_REQUESTS, kv::SecurityDomain::PUBLIC);
+  auto& pbft_pre_prepares = store->create<pbft::PrePrepares>(
+    pbft::Tables::PBFT_PRE_PREPARES, kv::SecurityDomain::PUBLIC);
+  auto replica_store = std::make_unique<pbft::Adaptor<ccf::Store>>(store);
+
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/test/replica_test.cpp]: Are we running any form of recovery for these tests? I think this can just be stubbed out.

File: src/consensus/pbft/pbft.h [hunk 13]
```diff
@@ -339,13 +345,36 @@ namespace pbft
...
+      for (auto&& [index, data, globally_committable] : entries)
+      {
+        replicate_to_ledger(data);
+      }
       return true;
     }
 
     bool replicate(const kv::BatchVector& entries) override
     {
+      for (auto&& [index, data, globally_committable] : entries)
```
Comment (ashamis) [inline src/consensus/pbft/pbft.h]: should this be a  universal reference rather than a normal reference?

File: src/consensus/pbft/pbftpreprepares.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  };
+
+  DECLARE_JSON_TYPE(PrePrepare);
+  DECLARE_JSON_REQUIRED_FIELDS(
+    PrePrepare, seqno, num_big_requests, digest_sig, contents);
+
+  // size_t is used as the key of the table. This key will always be 0 since we
+  // don't want to store the pre prepare in the kv over time, we just want to
+  // get them into the ledger
+  using PrePrepares = ccf::Store::Map<size_t, PrePrepare>;
```
Comment (ashamis) [inline src/consensus/pbft/pbftpreprepares.h]: the names are confusing, can you rename PrePrepares to something like PrePrepareMap?

File: src/consensus/pbft/pbftrequests.h [hunk 15]
```diff
@@ -63,11 +63,11 @@ namespace pbft
...
+  DECLARE_JSON_REQUIRED_FIELDS(Request, actor, caller_id, caller_cert, raw);
+
   // size_t is used as the key of the table. This key will always be 0 since we
   // don't want to store the requests in the kv over time, we just want to get
   // them into the ledger
-  using PbftRequests = ccf::Store::Map<size_t, Request>;
-
-  DECLARE_JSON_TYPE(Request);
-  DECLARE_JSON_REQUIRED_FIELDS(Request, actor, caller_id, caller_cert, raw);
+  using Requests = ccf::Store::Map<size_t, Request>;
```
Comment (ashamis) [inline src/consensus/pbft/pbftrequests.h]: same comment as the preprepares use something like RequestMap

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 16]
```diff
@@ -164,11 +168,26 @@ Replica::Replica(
...
+  ledger_writer = std::make_unique<LedgerWriter>(store, pbft_pre_prepares);
+  requests_local_hook = [this](
+                          kv::Version version,
+                          const pbft::Requests::State& s,
+                          const pbft::Requests::Write& w) {
+    for (auto& [key, value] : w)
+    {
+      playback_request(value.value);
+    }
+  };
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: Can this be just a function in the replica class?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 17]
```diff
@@ -164,11 +168,26 @@ Replica::Replica(
...
+
+  pre_prepares_local_hook = [this](
+                              kv::Version version,
+                              const pbft::PrePrepares::State& s,
+                              const pbft::PrePrepares::Write& w) {
+    for (auto& [key, value] : w)
+    {
+      playback_pre_prepare(value.value);
+    }
+  };
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: Same as above can this just be a function

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 18]
```diff
@@ -269,64 +288,60 @@ bool Replica::compare_execution_results(
...
 
-bool Replica::apply_ledger_data(const std::vector<uint8_t>& data)
+void Replica::playback_request(const pbft::Request& request)
 {
-  PBFT_ASSERT(ledger_replay, "ledger_replay should be initialized");
-
-  if (data.empty())
+  auto req =
+    create_message<Request>(request.raw.data(), request.raw.size()).release();
+  if (!(req->size() > Request::big_req_thresh && brt.add_request(req)))
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: i don't think you need the size check

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 19]
```diff
@@ -269,64 +288,60 @@ bool Replica::compare_execution_results(
...
+  ByzInfo info;
+  if (execute_tentative(executable_pp.get(), info))
   {
-    auto seqno = executable_pp->seqno();
-
-    ByzInfo info;
-    if (execute_tentative(executable_pp.get(), info))
+    auto batch_info = compare_execution_results(info, executable_pp.get());
+    if (!batch_info)
     {
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: you should probably write an error log line. At this point there is an attempted playback from a *bad* preprepare.....

File: src/consensus/pbft/libbyz/test/replica_main.cpp [hunk 20]
```diff
@@ -202,6 +203,14 @@ int main(int argc, char** argv)
...
   }
 
+  auto store = std::make_shared<ccf::Store>(
+    pbft::replicate_type_pbft, pbft::replicated_tables_pbft);
+  auto& pbft_requests = store->create<pbft::Requests>(
+    pbft::Tables::PBFT_REQUESTS, kv::SecurityDomain::PUBLIC);
+  auto& pbft_pre_prepares = store->create<pbft::PrePrepares>(
+    pbft::Tables::PBFT_PRE_PREPARES, kv::SecurityDomain::PUBLIC);
+  auto replica_store = std::make_unique<pbft::Adaptor<ccf::Store>>(store);
+
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/test/replica_main.cpp]: Are we running any form of recovery for these tests? I think this can just be stubbed out.

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 19 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Replica.cpp]: compare_execution_results will have already printed error messages if this returns false

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp [hunk 21]
```diff
@@ -107,24 +110,38 @@ TEST_CASE("Test Ledger Replay")
...
   ExecutionMock exec_mock(0);
 
-  // initiate replica with ledger enclave to be used on replay
-  ringbuffer::Circuit replay_circuit(circuit_size);
-  auto wf_rplay = ringbuffer::WriterFactory(replay_circuit);
-
-  auto replay_ledger_io = std::make_unique<consensus::LedgerEnclave>(wf_rplay);
-
-  create_replica(service_mem, std::move(replay_ledger_io));
+  auto store = std::make_shared<ccf::Store>(
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: It might be worth while leaving the comment to make this easier to read

File: src/consensus/pbft/libbyz/test/replica_main.cpp (hunk 20 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/replica_main.cpp]: No but we need to pass references to tables so it can't be nullptr here anymore. What do you mean stubbed out?

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 16 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Replica.cpp]: As in not be a local hook on a map? How will we know to call the functions after we call `Store::deserialise()`?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 22]
```diff
@@ -1075,6 +1101,18 @@ void Replica::emit_signature_on_next_pp(int64_t version)
   signed_version = version;
 }
 
+void Replica::activate_pbft_local_hooks()
+{
+  pbft_requests.set_local_hook(requests_local_hook);
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: can you add an assert that the hook is not set before overwriting it

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 16 above)
Comment (ashamis) [reply src/consensus/pbft/libbyz/Replica.cpp]: Do not create a hook here, Have a function that takes the same input and then loops over it an calls playback_request. Then when you set the hooks create the std::functions there. The std::functions do not need to be stored in replica object

File: src/consensus/pbft/libbyz/test/replica_main.cpp (hunk 20 above)
Comment (ashamis) [reply src/consensus/pbft/libbyz/test/replica_main.cpp]: have an abstract base class or adapter class that wraps the functionality that you need from the Table class. It would be good to reduce coupling if we can.

File: src/kv/test/kv_serialisation.cpp [hunk 23]
```diff
@@ -470,10 +470,10 @@ TEST_CASE("replicated and derived table serialisation")
       data_replicated_private,
       data_derived,
       data_derived_private);
-    data_view_r->put(42, 42);
-    data_view_d->put(42, 42);
-    data_view_r_p->put(43, 43);
-    data_view_d_p->put(43, 43);
+    data_view_r->put(44, 44);
```
Comment (jumaffre) [inline src/kv/test/kv_serialisation.cpp]: Why these changes?

File: src/consensus/pbft/pbfttables.h [hunk 24]
```diff
@@ -9,9 +9,10 @@ namespace pbft
   struct Tables
```
Comment (jumaffre) [inline src/consensus/pbft/pbfttables.h]: Minor: `#pragma once` missing at the top of the file.

File: src/consensus/pbft/libbyz/LedgerWriter.cpp [hunk 25]
```diff
@@ -42,73 +44,17 @@ void LedgerWriter::write_prepare(
       (uint8_t*)&(p.second),
       sizeof(Prepared_cert::PrePrepareProof));
   }
-
-  ledger->put_entry(entry);
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/LedgerWriter.cpp]: Should this be removed? It seems that now, `write_prepare()` has no effect?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 26]
```diff
@@ -1075,6 +1101,18 @@ void Replica::emit_signature_on_next_pp(int64_t version)
...
+  pbft_pre_prepares.set_local_hook(pre_prepares_local_hook);
+}
+
+void Replica::deactivate_pbft_local_hooks()
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Replica.cpp]: Unused?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 27]
```diff
@@ -1075,6 +1101,18 @@ void Replica::emit_signature_on_next_pp(int64_t version)
   signed_version = version;
 }
 
+void Replica::activate_pbft_local_hooks()
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Replica.cpp]: Question: why is this required? It seems that it is only used in the unit test.

File: src/consensus/pbft/libbyz/LedgerWriter.h [hunk 28]
```diff
@@ -4,22 +4,23 @@
...
 class LedgerWriter
 {
 private:
-  std::unique_ptr<consensus::LedgerEnclave> ledger;
+  pbft::Store& store;
+  pbft::PrePrepares& pbft_pre_prepares;
 
 public:
-  LedgerWriter(std::unique_ptr<consensus::LedgerEnclave> ledger_);
+  LedgerWriter(pbft::Store& store_, pbft::PrePrepares& pbft_pre_prepares_);
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/LedgerWriter.h]: I think it would be better to pass a reference to `PrePrepares` to the `pbft::Adaptor` in `nodestate.h` instead of passing it through here and `Replica`?

File: src/kv/test/stub_consensus.h [hunk 29]
```diff
@@ -46,6 +46,20 @@ namespace kv
         return std::make_pair(std::vector<uint8_t>(), false);
     }
 
+    std::pair<std::vector<uint8_t>, bool> replay_data()
```
Comment (jumaffre) [inline src/kv/test/stub_consensus.h]: Better to call this `pop_oldest_data()`?

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp [hunk 30]
```diff
@@ -166,124 +198,68 @@ TEST_CASE("Test Ledger Replay")
...
+        seqno++;
       }
       else
       {
-        last = initial_ledger.begin() + positions.at(until);
+        // even entries are requests
+        auto req_view = tx.get_view(pbft_requests);
+        auto req = req_view->get(0);
+        REQUIRE(req.has_value());
+        REQUIRE(req.value().raw.size() > 0);
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: I am not sure these two lines test much? Since the requests are always written at index `0`, this test only asserts that a request was written _at some point_ but not necessarily in the latest iteration.

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp [hunk 31]
```diff
@@ -145,6 +162,21 @@ TEST_CASE("Test Ledger Replay")
...
+         {},
+         {(const uint8_t*)request->contents(),
+          (const uint8_t*)request->contents() + request->size()}});
+
+      auto der_view = tx.get_view(write_derived_map);
+      der_view->put("key1", "value1");
+
+      REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+
       // imitate exec command
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: How difficult would it be to actually have this work done in `ExecutionMock`'s `exec_command()`? IMO, it would give a much more accurate test (i.e. actually relying on the calls to `ledger_writer->write_pre_prepare()` in `Replica.cpp`), instead of having to play with the KV's interface.

File: src/consensus/pbft/pbft.h [hunk 32]
```diff
@@ -339,13 +345,36 @@ namespace pbft
...
+    }
+
+    template <>
+    size_t replicate_to_ledger<std::vector<uint8_t>>(
+      const std::vector<uint8_t>& data)
+    {
+      ledger->put_entry(data);
+      return data.size();
+    }
+
```
Comment (jumaffre) [inline src/consensus/pbft/pbft.h]: Same as in `raft.h`.

File: src/consensus/pbft/pbft.h [hunk 33]
```diff
@@ -339,13 +345,36 @@ namespace pbft
...
     bool replicate(const kv::BatchDetachedBuffer& entries) override
     {
+      for (auto&& [index, data, globally_committable] : entries)
+      {
+        replicate_to_ledger(data);
```
Comment (jumaffre) [inline src/consensus/pbft/pbft.h]: Minor: should this be caller `store_in_ledger()` or similar as this does not actually replicate (as opposed to raft's replicate).

File: src/consensus/pbft/pbfttables.h (hunk 24 above)
Comment (jumaffre) [reply src/consensus/pbft/pbfttables.h]: Addressed in https://github.com/microsoft/CCF/pull/663

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 19 above)
Comment (jumaffre) [reply src/consensus/pbft/libbyz/Replica.cpp]: Question: does the replica recover gracefully if something goes wrong here?

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 34]
```diff
@@ -269,64 +288,60 @@ bool Replica::compare_execution_results(
...
 
-  for (auto& executable_pp : executable_pps)
+  ByzInfo info;
+  if (execute_tentative(executable_pp.get(), info))
   {
-    auto seqno = executable_pp->seqno();
-
-    ByzInfo info;
-    if (execute_tentative(executable_pp.get(), info))
+    auto batch_info = compare_execution_results(info, executable_pp.get());
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Replica.cpp]: Minor: this can be `if(compare_execution_results(info, executable_pp.get()) { return; }`.

File: src/consensus/pbft/libbyz/LedgerWriter.cpp (hunk 25 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/LedgerWriter.cpp]: Yes all things manually writing to the ledger will be removed in next PR

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 26 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Replica.cpp]: To be used in next PR where playback of ledger will happen via State messages

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 27 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Replica.cpp]: Will be used in next PR where playback will happen via State messages. When we are not playing back from the ledger we don’t want the hooks to be triggered but we want execution to go on as usual

File: src/consensus/pbft/libbyz/Replica.cpp (hunk 19 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Replica.cpp]: No we need to implement rollbacks #457

File: src/consensus/pbft/libbyz/LedgerWriter.h (hunk 28 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/LedgerWriter.h]: `Replica` needs the tables anyway to set the hooks so it would make sense to pass the table here too. Right now the `pbft::Adaptor` doesn't store any state apart from the `Store` so I am not sure I would like to have the `Adaptor` keep a reference to a table.

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 31 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: Not sure this has to do with work done in `exec_command`. to have `Replica` call `ledger_writer->write_pre_prepare()` we need a full run of the protocol where requests are sent and are processed via pre-prepare, prepares, etc (like the `test_UDP` end to end test for example). Here we just want to populate the ledger with requests and pre-prepares so that we unit test playing back the ledger further down.

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 30 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: This is mostly here to make sure that the replicated data got deserialised and the derived data didn't. We could potentially add another check on the actual request to make sure that everything is being deserialised correctly. The main check is done further down where we see if `last_executed == total_requests -1` but I'll see if I can get more useful checking here

File: src/consensus/pbft/pbftpreprepares.h (hunk 14 above)
Comment (olgavrou) [reply src/consensus/pbft/pbftpreprepares.h]: Renaming `pbft::PrePrepares` to `pbft::PrePreparesMap`, the table `pbft_pre_prepares` to `pbft_pre_prepares_map` and renaming `pbft::Requests` to `pbft::RequestsMap`, and the table `pbft_requests` to `pbft_requests_map`

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 31 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: Changing to trigger `ledger_writer->write_pre_prepare()` via `Replica::handle(Request)` as discussed, instead of manually creating ad writing the pre-prepares.

File: src/consensus/pbft/libbyz/test/test_ledger_replay.cpp (hunk 30 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/test/test_ledger_replay.cpp]: added extra check in exec_command to make sure requests are getting executed in order

File: src/consensus/pbft/libbyz/parameters.h (hunk 10 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/parameters.h]: https://github.com/microsoft/CCF/issues/671

**Comment by cimetrics** [comment]
store_pre_prepares@[3646 aka 20200106.7](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3646&view=results) vs master ewma over 20 builds from [3569](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3569&view=results) to [3644](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3644&view=results)
images

**Comment by cimetrics** [comment]
store_pre_prepares@[3648 aka 20200106.8](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3648&view=results) vs master ewma over 20 builds from [3569](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3569&view=results) to [3644](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3644&view=results)
images

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 35]
```diff
@@ -1035,6 +1030,17 @@ void Replica::register_global_commit(global_commit_handler_cb cb, void* ctx)
   global_commit_ctx = ctx;
 }
 
+template <class T>
+std::unique_ptr<T> Replica::create_message(
+  const uint8_t* message_data, size_t data_size)
+{
+  Message* m = new Message(data_size);
+  memcpy(m->contents(), message_data, data_size);
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Replica.cpp]: std::copy?

**Comment by cimetrics** [comment]
store_pre_prepares@[3658 aka 20200106.13](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3658&view=results) vs master ewma over 10 builds from [3612](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3612&view=results) to [3656](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3656&view=results)
images

**Comment by cimetrics** [comment]
store_pre_prepares@[3665 aka 20200107.3](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3665&view=results) vs master ewma over 20 builds from [3589](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3589&view=results) to [3663](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3663&view=results)
images

**Comment by cimetrics** [comment]
store_pre_prepares@[3667 aka 20200107.4](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3667&view=results) vs master ewma over 20 builds from [3589](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3589&view=results) to [3663](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3663&view=results)
images
